### ISSUE TITLE
Error when trying execute a shell script

### ISSUE THREAD

**Issue by brenogil** [fish-shell/fish-shell#2783]
I have recently switched to fish and I have been trying to run some .sh scripts. 
Usually I have no problem but for one particular script I get the error below.
### Observed behavior:

```
Failed to execute process './alt_script.sh'. Reason:
', which is not an executable command.he interpreter '/bin/sh
```

I believe at least the message could be improved, if nothing else. 
### Other comments:

I have tried playing with different shebangs:
# !/bin/sh
# !/bin/bash
# !/usr/local/bin/fish

---

Fish version: fish, version 2.2.0

Operating system: OSX 10.10.5 - Homebrew

**Comment by faho** [comment]
Could you upload the script?

>', which is not an executable command.he interpreter '/bin/sh

I do believe you've made a copy-paste mistake here - for me the error is

`The file 'FILENAME' specified the interpreter 'INTERPRETER', which is not an executable command`.

Also, check that the mentioned interpreter actually exists and is executable (`test -x $INTERPRETER`).

And what is your locale (`echo $LANG`, dunno if OSX has a `locale` command)?

**Comment by brenogil** [comment]
> I do believe you've made a copy-paste mistake here 

![screen shot 2016-03-01 at 17 17 43](https://cloud.githubusercontent.com/assets/5962893/13431300/8dae3500-dfd1-11e5-98e1-3e8e62a1eb6d.png)

> Also, check that the mentioned interpreter actually exists and is executable (test -x $INTERPRETER).
 
Returns nothing. Could this be the problem?

> And what is your locale (echo $LANG, dunno if OSX has a locale command)?

locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL="en_US.UTF-8"

**Comment by faho** [comment]
>Returns nothing. Could this be the problem?

It should set $status to 1 if it doesn't - your prompt displays a red cross then if I'm not mistaken.

> LC_ALL="en_US.UTF-8"

You usually do not want to set LC_ALL, as it overrides all other locale variables.

Otherwise, I got nothing. No clue - anyone else? @fish-shell/contributors?

**Comment by brenogil** [comment]
> You usually do not want to set LC_ALL, as it overrides all other locale variables.
Ok. I undid it. 

> It should set $status to 1 if it doesn't - your prompt displays a red cross then if I'm not mistaken.
... I'm not sure I fully understand this. Are there any resources about it around?

**Comment by faho** [comment]
http://fishshell.com/docs/current/index.html#variables (a bit down, under "The status variable"):

>Whenever a process exits, an exit status is returned to the program that started it (usually the shell). This exit status is an integer number, which tells the calling application how the execution of the command went. In general, a zero exit status means that the command executed without problem, but a non-zero exit status means there was some form of problem. Fish stores the exit status of the last process in the last job to exit in the `status` variable.

and

http://fishshell.com/docs/current/tutorial.html#tut_exit_status.

Basically, `test` returns a status. When the test fails (i.e. the file isn't executable - which it also can't be if it doesn't exist) it will return 1. Your prompt acts on it if the last command returned something > 0 and displays a red cross, like you can see in your screenshot, so you can just execute a test and see if it failed.

**Comment by brenogil** [comment]
Thanks! 
test -x alt_script.sh
test alt_script.sh
test -x $INTERPRETER

all return $status 0

**Comment by faho** [comment]
> test alt_script.sh

That's not how test works. It has a few operators, but if you're not handing it any, it will check if the arguments are not empty (in a pure text sense).See http://fishshell.com/docs/current/commands.html#test for more information.

> test -x $INTERPRETER

By "$INTERPRETER", I meant the full path to the interpreter you're using in the shebang, e.g. `test -x /bin/bash` (that test returns 0 in this case is #2037, an instance of POSIX, the standard we follow for it, being stupid).

**Comment by brenogil** [comment]
stupid me! 

test -x /bin/bash
test -x /bin/sh

all return 0. 

As I said, I have been able to run similar scripts. 
I find the error message strange more than anything

**Comment by faho** [comment]
>As I said, I have been able to run similar scripts. 

Which is why the question is: What's different about this one? Can you upload it?

**Comment by faho** [comment]
Did this script ever have contact with a windows system? Does it work if you copy it by hand (i.e. not copy-and-pasting, but entering the text anew in a new file)?

**Comment by faho** [comment]
I can reproduce the weird error message with the following script:

```fish
printf '%s\r' \#!/bin/bash 'echo banana' > why.sh
chmod +x why.sh
./why.sh
```

You might notice that printf here has a "\r" instead of the more usual "\n". That's what windows usually uses for line endings, and it's been a pain for the last two decades.

To fix your script, run `sed 's/\r/\n' -i /path/to/script` or `dos2unix` or something like that.

It's possible we could detect that and print a better error message.

**Comment by faho** [comment]
Or, since you use emacs, open it with that and do M-x set-buffer-file-coding-system<RET> utf-8-unix, then save.

**Comment by brenogil** [comment]
>Did this script ever have contact with a windows system? 
Yes! It is a script I tried to convert from a .bat

>Or, since you use emacs, open it with that and do M-x set-buffer-file-coding-system utf-8-unix, then save.
This worked in the sense that error message does not appear anymore. 
./alt_script.sh: line 20: wine: command not found
./alt_script.sh: line 24: wine: command not found
./alt_script.sh: line 29: wine: command not found
./alt_script.sh: line 34: wine: command not found

>Does it work if you copy it by hand (i.e. not copy-and-pasting, but entering the text anew in a new file)?
Yes

**Comment by faho** [comment]
> ./alt_script.sh: line 20: wine: command not found

Let me go out on a limb here - my guess is, and this is purely a guess, that you either don't have wine installed or don't have it installed in $PATH (`command -s wine` doesn't work).

**Comment by brenogil** [comment]
I don't think its is the case. I've been relying heavily on wine. (that sounds bad)
./alt_script.sh: line 52: mkdir: command not found
./alt_script.sh: line 64: rm: command not found
./alt_script2.sh: line 66: `done'

I guess I'll have to figure it out what is going on...

**Comment by faho** [comment]
If you want, you can upload the script and I'll have a look. That it doesn't find `mkdir` or `rm` sounds like you misconfigured $PATH (your config.fish would be nice as well, my guess is you have something like `set PATH /usr/local/bin:$PATH` in there when it should be `set PATH /usr/local/bin $PATH` - note the colon). That it doesn't find `done` is to be expected, just like it wouldn't find `fi` or `esac` - all those are `end` in fish.

**Comment by brenogil** [comment]
[alt_script.sh.zip](https://github.com/fish-shell/fish-shell/files/153233/alt_script.sh.zip)
This is the file.

It is a direct copy from this one

[different_ways_to_create_a_chm.bat.zip](https://github.com/fish-shell/fish-shell/files/153234/different_ways_to_create_a_chm.bat.zip)

**Comment by faho** [comment]
The last `done` is wrong, because you've not started a for/while loop - just remove it.

Otherwise it looks okay.

Add `echo $PATH` to the top to check what sh thinks your path is, and then confirm that `rm` and `mkdir` exist in one of those directories.

**Comment by brenogil** [comment]
Thanks!
I'd say most of it was outside of the actual fish scope, but I'd be glad if at least this effort improves that error message a bit. 
:)

**Comment by faho** [comment]
Okay then. If you have any other questions, feel free to ask!

**Comment by Optiligence** [comment]
in summary: the problem is windows line endings in the script

>PS/my 2 cents: i think it’s odd that this just breaks

**Comment by endolith** [comment]
It should just handle all line endings correctly.  It produces weird error messages:


![image](https://user-images.githubusercontent.com/58611/66234479-ccc13e00-e6bb-11e9-8292-c617f8ec819d.png)

**Comment by faho** [comment]
@endolith Tell that to the kernel folks, which is where the error comes from. Fish has nothing to do with this and can't do anything about it.

**Comment by zanchey** [comment]
The error message is super weird though! At least if it was "The file './build.sh' specified the interpreter '/usr/bin/fish\r', which is not an executable command.", people would have some hope of working it out.

**Comment by lilyball** [comment]
It's super weird because of the raw control character in there (the carriage return). The kernel isn't doing any escaping of control characters in its output.

**Comment by zanchey** [comment]
Right. Unfortunately doing that escaping safely in postfork is a bit irritating; I have a hacky patch but it's probably not the right approach.

**Comment by zanchey** [comment]
Reopening this to track the improvement.

**Comment by faho** [comment]
Okay, I added a change that detects windows line endings in the shebang and complains about them.

That should fix, by my estimate, 99.7% of the problem.
